File: Sources/CourseReader/Services/QuizEngine.swift
import Foundation

@MainActor
final class QuizEngine: ObservableObject {
  @Published var questions: [QuizQuestion] = []
  @Published var currentIndex = 0
  @Published var selectedAnswers: [String: String] = [:]
  @Published var showResults = false
  @Published var isCompleted = false

  var currentQuestion: QuizQuestion? {
    guard currentIndex < questions.count else { return nil }
    return questions[currentIndex]
  }

  var score: (correct: Int, total: Int) {
    let total = questions.count
    let correct = questions.filter { selectedAnswers[$0.id] == $0.correctOption }.count
    return (correct, total)
  }

  var percentage: Double {
    guard score.total > 0 else { return 0 }
    return Double(score.correct) / Double(score.total) * 100
  }

  func load(_ qs: [QuizQuestion]) {
    questions = qs
    currentIndex = 0
    selectedAnswers = [:]
    showResults = false
    isCompleted = false
  }

  func selectAnswer(_ answer: String) {
    guard let q = currentQuestion else { return }
    selectedAnswers[q.id] = answer
  }

  func nextQuestion() {
    guard currentIndex < questions.count - 1 else {
      isCompleted = true
      showResults = true
      return
    }
    currentIndex += 1
  }

  func isCorrect(_ questionId: String) -> Bool? {
    guard let selected = selectedAnswers[questionId],
          let q = questions.first(where: { $0.id == questionId })
    else { return nil }
    return selected == q.correctOption
  }

  func reset() {
    questions = []
    currentIndex = 0
    selectedAnswers = [:]
    showResults = false
    isCompleted = false
  }
}

File: src/bun/quiz-engine.ts
import type { QuizQuestion } from './types';

export class QuizEngine {
  questions: QuizQuestion[] = [];
  currentIndex = 0;
  selectedAnswers: Record<string, string> = {};
  isCompleted = false;
  courseId?: string;
  moduleId?: string;

  get currentQuestion(): QuizQuestion | null {
    return this.currentIndex < this.questions.length ? this.questions[this.currentIndex] : null;
  }

  get score(): { correct: number; total: number } {
    const total = this.questions.length;
    const correct = this.questions.filter((q) => this.selectedAnswers[q.id] === q.answer).length;
    return { correct, total };
  }

  get percentage(): number {
    return this.score.total > 0 ? (this.score.correct / this.score.total) * 100 : 0;
  }

  load(qs: QuizQuestion[], courseId?: string, moduleId?: string): void {
    this.questions = qs;
    this.currentIndex = 0;
    this.selectedAnswers = {};
    this.isCompleted = false;
    this.courseId = courseId;
    this.moduleId = moduleId;
  }

  selectAnswer(answer: string): void {
    const q = this.currentQuestion;
    if (!q) return;
    this.selectedAnswers[q.id] = answer;
  }

  nextQuestion(): void {
    if (this.currentIndex < this.questions.length - 1) {
      this.currentIndex += 1;
    } else {
      this.isCompleted = true;
    }
  }

  isCorrect(questionId: string): boolean | null {
    const selected = this.selectedAnswers[questionId];
    const q = this.questions.find((q) => q.id === questionId);
    if (!selected || !q) return null;
    return selected === q.answer;
  }

  reset(): void {
    this.questions = [];
    this.currentIndex = 0;
    this.selectedAnswers = {};
    this.isCompleted = false;
  }
}
